test/Microsoft.AspNet.Mvc.FunctionalTests/Logging/LoggingExtensions.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using LoggingWebSite;
+using Microsoft.AspNet.WebUtilities;
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public static class LoggingExtensions
+    {
+        public const string RequestTraceIdQueryKey = "RequestTraceId";
+
+        public static IEnumerable<LogInfoDto> GetStartupLogs(this IEnumerable<ActivityContextDto> activities)
+        {
+            return activities.Where(activity => activity.RequestInfo == null)
+                             .GetLogs();
+        }
+
+        public static IEnumerable<LogInfoDto> GetStartupLogs(this IEnumerable<ActivityContextDto> activities,
+                                                            string scopeName)
+        {
+            return activities.Where(activity => activity.RequestInfo == null)
+                             .GetLogsUnderScope(scopeName);
+        }
+
+        public static IEnumerable<LogInfoDto> GetLogs(this IEnumerable<ActivityContextDto> activities,
+                                                     string requestTraceId)
+        {
+            return activities.FilterByRequestTraceId(requestTraceId)
+                             .GetLogs();
+        }
+
+        public static IEnumerable<LogInfoDto> GetLogs(this IEnumerable<ActivityContextDto> activities,
+                                                     string requestTraceId,
+                                                     string scopeName)
+        {
+            return activities.FilterByRequestTraceId(requestTraceId)
+                             .GetLogsUnderScope(scopeName);
+        }
+
+        private static IEnumerable<LogInfoDto> GetLogsUnderScope(this IEnumerable<ActivityContextDto> activities,
+                                                        string scopeName)
+        {
+            var logInfos = new List<LogInfoDto>();
+
+            foreach (var activity in activities)
+            {
+                if (activity.RepresentsScope)
+                {
+                    GetLogsUnderScopeHelper(activity.Root, logInfos, scopeName, foundScope: false);
+                }
+            }
+
+            return logInfos;
+        }
+
+        private static IEnumerable<LogInfoDto> GetLogs(this IEnumerable<ActivityContextDto> activities)
+        {
+            // Build a flat list of log messages from the log node tree 
```

Comment:
> we would have to however traverse through the entire(worst case scenario) tree structure to find the log message having this structure type...

If we find a performance issue in the test code then we'll address it :laughing: Correctness and depth wins every time in tests